### PR TITLE
resolve trivial merge of `A+A-B+C-C` to `A`

### PR DESCRIPTION
**Pull request by martinvonz**
# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (src/config-schema.json)
- [ ] I have added tests to cover my changes

### CODE REVIEW

**Comment by ilyagr** [comment]
I am not actually sure about this (nor am I sure this is bad). I'd be OK with having two degrees of trivializing conflicts: `trivial_merge` and `simplify(trivial_merge)` that are different. We'd pick which to use on a case-by-case basis (though I'm not sure how exactly to make that decision).

The argument for this change is pretty self-evident.

The main argument against this change, in my mind, is the ambiguous conflict we discussed on Discord (last discussion [here](https://discord.com/channels/968932220549103686/969291218347524238/1107100739916009562)). It resolves differently depending on which side you resolve first:

```
N   X   X
 \ / \ /
  X   N
```

With this change, I think it will just resolve to `X`. I'm not sure if that's OK.

-----

In any case, I think one of us should add a test for this conflict specifically.

**Comment by martinvonz** [comment]
> I am not actually sure about this (nor am I sure this is bad). I'd be OK with having two degrees of trivializing conflicts: `trivial_merge` and `simplify(trivial_merge)` that are different. We'd pick which to use on a case-by-case basis (though I'm not sure how exactly to make that decision).

Yes, we may want to do that. I think that's basically what the TODO I moved in this PR is about, right?

> The main argument against this change, in my mind, is the ambiguous conflict we discussed on Discord (last discussion [here](https://discord.com/channels/968932220549103686/969291218347524238/1107100739916009562)). It resolves differently depending on which side you resolve first:
> 
> ```
> N   X   X
>  \ / \ /
>   X   N
> ```
> 
> With this change, I think it will just resolve to `X`. I'm not sure whether that's OK in every case (in contexts like resolving executable bits, it's probably fine).

I'm pretty confident that that it's the resolution of the form `A+A-B => A` that's the issue there. I think it would be more correct to what Darcs and Pijul do there, but I also think it would be more annoying for the user. I think we talked about possibly doing that kind of resolution only in some contexts (that's the TODO). For example, we could decide to leave such conflicts unresolved until the user does `jj resolve`. However, I suspect way more users would complain about that the number of users who ran into the corner case where the stricter merging saved them.

> In any case, I think one of us should add a test for this conflict example specifically.

I think that's what https://github.com/martinvonz/jj/blob/d13db728ba8d13ef9e67afb1b3b09cc6cfc4abe0/lib/src/merge.rs#LL111C1-L112C1 (and lines 118 and 120) do, isn't it? It's not specific to the executable bit, but given that we have pretty thorough unit tests of `trivial_merge()` and that we use that function for resolving the executable bit, I don't think we need to test many cases in the e2e tests. What do you think?

File: lib/src/merge.rs [hunk 1]
```diff
@@ -92,48 +98,48 @@ mod tests {
...
+        assert_eq!(trivial_merge(&[0, 1], &[0, 2, 2]), Some(&2));
         assert_eq!(trivial_merge(&[0, 1], &[0, 2, 3]), None);
         assert_eq!(trivial_merge(&[0, 1], &[1, 0, 0]), Some(&0));
         assert_eq!(trivial_merge(&[0, 1], &[1, 0, 1]), Some(&1));
         assert_eq!(trivial_merge(&[0, 1], &[1, 0, 2]), Some(&2));
         assert_eq!(trivial_merge(&[0, 1], &[1, 1, 0]), Some(&1));
-        assert_eq!(trivial_merge(&[0, 1], &[1, 1, 1]), None);
+        assert_eq!(trivial_merge(&[0, 1], &[1, 1, 1]), Some(&1));
         assert_eq!(trivial_merge(&[0, 1], &[1, 1, 2]), None);
         assert_eq!(trivial_merge(&[0, 1], &[1, 2, 0]), Some(&2));
```
Comment (ilyagr) [inline lib/src/merge.rs:129]: This has little to do with this PR, but I find this example problematic:

```
1    2    0
 ^   ^   ^
  \ / \ /
   0   1
```

Is it OK to be this cavalier with the ordering? If we respect it, the resolution is clearly ambiguous.

Perhaps we should leave a note about it as well.

**Comment by ilyagr** [comment]
I looked through the examples, and concluded that most of my concerns have little to do with this PR. The changes caused specifically by this PR all seem to be for the better. Reviewing it did raise a bunch of questions in my mind, though.

> > In any case, I think one of us should add a test for this conflict example specifically.
> 
> I think that's what https://github.com/martinvonz/jj/blob/d13db728ba8d13ef9e67afb1b3b09cc6cfc4abe0/lib/src/merge.rs#LL111C1-L112C1 (and lines 118 and 120) do ..., isn't it? ... I don't think we need to test many cases in the e2e tests. What do you think?

Absolutely, I think what you have is sufficient, I just didn't look carefully enough. In fact, the tests you made for the trivial merge are really excellent.

As that test shows, this was an issue before this PR. That's a bit concerning to me, but it clearly is a separate problem. I wrote more next to that line.

(My concern is limited, though, by the fact that in 90% of the cases resolving conflicts one 2-add conflict at a time is the best way, and there are no ambiguities there).

> I'm pretty confident that that it's the resolution of the form `A+A-B => A` that's the issue there. I think it would be more correct to what Darcs and Pijul do there, but I also think it would be more annoying for the user. I think we talked about possibly doing that kind of resolution only in some contexts (that's the TODO). For example, we could decide to leave such conflicts unresolved until the user does `jj resolve`. However, I suspect way more users would complain about that the number of users who ran into the corner case where the stricter merging saved them.

I agree about the user experience. It seems to force us into accepting the `A+A-B => A` rule, but that means we also have to accept the ambiguities it causes. Hopefully, we'll come up with a clearer way of understanding some logic behind these, but I don't think we have done that yet.

> > I am not actually sure about this (nor am I sure this is bad). I'd be OK with having two degrees of trivializing conflicts: `trivial_merge` and `simplify(trivial_merge)` that are different. We'd pick which to use on a case-by-case basis (though I'm not sure how exactly to make that decision).
> 
> Yes, we may want to do that. I think that's basically what the TODO I moved in this PR is about, right?

That TODO (if I looked at the right one) is strictly about the conflicts where the adds are the same. There are also other simplifications that are questionable, especially ones that reorder the adds in an arbitrary way. Even this PR is a bit questionable theoretically, even though in practice it seems to only make things better.

**Review by ilyagr — APPROVED** [review]

File: lib/src/merge.rs [hunk 2]
```diff
@@ -92,48 +98,48 @@ mod tests {
...
         assert_eq!(trivial_merge(&[0, 1], &[1, 2, 1]), None);
-        assert_eq!(trivial_merge(&[0, 1], &[1, 2, 2]), None);
+        assert_eq!(trivial_merge(&[0, 1], &[1, 2, 2]), Some(&2));
```
[the remark this answers is not in the document]
Comment (ilyagr) [reply lib/src/merge.rs:131]: by "has little to do with this PR", I mean that the root cause seems to be the reordering, similar to https://github.com/martinvonz/jj/pull/1697#discussion_r1230233419.

File: lib/src/merge.rs [hunk 3]
```diff
@@ -92,48 +98,48 @@ mod tests {
...
         assert_eq!(trivial_merge(&[0, 1], &[0, 1, 2]), Some(&2));
         assert_eq!(trivial_merge(&[0, 1], &[0, 2, 0]), None);
         assert_eq!(trivial_merge(&[0, 1], &[0, 2, 1]), Some(&2));
-        assert_eq!(trivial_merge(&[0, 1], &[0, 2, 2]), None);
+        assert_eq!(trivial_merge(&[0, 1], &[0, 2, 2]), Some(&2));
         assert_eq!(trivial_merge(&[0, 1], &[0, 2, 3]), None);
         assert_eq!(trivial_merge(&[0, 1], &[1, 0, 0]), Some(&0));
         assert_eq!(trivial_merge(&[0, 1], &[1, 0, 1]), Some(&1));
         assert_eq!(trivial_merge(&[0, 1], &[1, 0, 2]), Some(&2));
         assert_eq!(trivial_merge(&[0, 1], &[1, 1, 0]), Some(&1));
```
[the remark this answers is not in the document]
Comment (ilyagr) [reply lib/src/merge.rs:126]: By the way, this is also an ordering issue. It's pretty clear that

```
1    1    0
 ^   ^   ^
  \ / \ /
   1   0
```

should resolve to 1.

File: lib/src/merge.rs (hunk 1 above)
Comment (ilyagr) [reply lib/src/merge.rs:129]: OTOH, if we think about this as a sequence of diffs, `1 0->2 1->0`, it seems less absurd to reorder them.

Comment (martinvonz) [reply lib/src/merge.rs:129]: Yes, I think of these as diffs.

Let's say the user merged commits A and B with common ancestor O and resolved a conflict, resulting in commit C. The auto-merged parent state would then be `A+(B-O)` (assuming we treat the first add as the base to apply diffs onto). If the user then does `jj rebase -r C -d B -d A` (i.e. rebases the resolution onto a merge done in the opposite order), we would get the new state `B+(A-O)+(C-(A+(B-O)))` = `B+A-O+C-A-B+O`, or `removes = [O,A,B], adds=[B,A,C,O]`. If we don't allow reordering (i.e. we only allow canceling the first remove against either of the first two adds?), then we wouldn't be able to resolve this, right?

Comment (martinvonz) [reply lib/src/merge.rs:129]: The rest of your comments seem to depend on what we decide about ordering, so I'll await your reply to this comment before I address (or not) the other comments.

File: lib/src/merge.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ilyagr) [reply lib/src/merge.rs:126]: On third thought, leaving such a conflict ambiguous would do no good. I think the fundamental problem here is as follows, and I don't know how we can resolve it and keep the `A+A-B => A` rule.

I think it would be nice to document this somewhere, but it doesn't have to be in this PR. Also, I'm not sure whether I'll have a better idea of what to say in the future.

--------------------------------

Let's say this is now a portion of a commit tree and we name all the changes:

```
A    B    C
1    1    0
     ^   ^
    /   /
   0   1
   D   E
```

The real problem is that if we do `jj rebase -s C -d B && jj amend -r C && jj rebase -s B -d A`, you get a different result than if you do `jj rebase -s C -d B && jj rebase -s B -d A && jj amend -r C`.

After the first rebase, the value of C is still 0, so you get

```
1
A

0 -> 1 -> 0
D    B    C

1
E
```

#### First option

Now, if we rebase `B` onto `A`, `B` remains a `1` by the `A+A-B => A`. So, `C` remains a 0. In other words, we get

```
1 -> 1 -> 0
A    B    C
```

and after the squash we have

```
1 -> 0
A    B and C squashed
```

#### Second option
If we `jj amend -r C` first, we end up with

```
1
A

0 -> 0
D    B and C squashed

1
E
```

At this point, `B` does not change the value, so rebasing it on top of `A` will result in

```
1 -> 1
A    B and C squashed
```

This is clearly not great, but I have no clue how to do any better since losing the `A+A-B => A` rule is so inconvenient.

File: lib/src/merge.rs (hunk 1 above)
Comment (ilyagr) [reply lib/src/merge.rs:129]: I don't have any actionable comments. The issue with ordering comes from a previous PR. Ideally, I'd like to include an explanation of what the other options are in comments, but I'm not sure my thoughts are clear enough for that at the moment.

Here are some non-actionable thoughts.

> If we don't allow reordering (i.e. we only allow canceling the first remove against either of the first two adds?), then we wouldn't be able to resolve this, right?

I think that's right, and I think there might be some benefit in theory to not allowing that. That would allow us to preserve the distinction of "our" side and "their" side when rebasing.

Another argument for allowing reordering is something we advertise in our README. Let's say we have a commit chain `A->B->C` and we reorder B and C to get `A->C'->B'`. As a conflict, I think `B'` can be written as `A B->C A->B`. `jj` declares that `B'` should *not* be conflicted (this is what we advertise in the README) and be equivalent to `C` .  

This suggests that we might want to allow reordering that preserves the diff structure, e.g. we'd like `A B->C D->E` to be the same as `A D->E B->C` even if we differentiate it from `A B->E D->C`. In principle, this is more restrictive than the arbitrary kind of reordering we currently allow. In terms of this PR, however, this more restrictive kind of reordering seems sufficient to resolve every conflicts I was concerned about.

I think it's possible we'll come up with use-cases for preserving some (e.g. up to diff structure) or all order info, but it seems we haven't gotten there yet.

File: lib/src/merge.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ilyagr) [reply lib/src/merge.rs:126]: One potential conclusion from this is that it might end up being useful to keep track which conflict resolutions rely on the `A+A-B => A` rule and which don't. At the moment, that seems more important to me than the discussion about reordering, but even so -- I currently have no use-case for doing it. Keeping track of such information is also not trivial theoretically -- we'd need a model for a conflict that tracks this information in a useful way and can still be rebased on top of any other such conflict in a meaningful way.

File: lib/src/merge.rs (hunk 1 above)
Comment (martinvonz) [reply lib/src/merge.rs:129]: Yeah, I have also felt that it's a bit weird to reorder `A B->C D->E` to `A B->E D->C`. Doing that means that we may produce diffs that did not appear in a diff compared to a parent. I'll work on a patch to change `Conflict::simplify()` to be more careful with how it reorders. I'll merge this as is since it doesn't change anything about ordering.

Comment (ilyagr) [reply lib/src/merge.rs:129]: As I said, in all of your examples, it doesn't seem to matter in the end, at least as far as `simplify` goes. This was surprising and weird to me. I suspect it would make a difference in larger conflicts, but they are not common.

Comment (ilyagr) [reply lib/src/merge.rs:129]: Also, I don't know if I made this clear: in my mind swapping the order of a merge is closely related to reordering `A B->C D->E` to `A B->E D->C`. That's another reason I hesitate about that in theory.

Comment (martinvonz) [reply lib/src/merge.rs:129]: > Also, I don't know if I made this clear: in my mind swapping the order of a merge is closely related to reordering `A B->C D->E` to `A B->E D->C`. That's another reason I hesitate about that in theory.

I can at least see how it's like reordering from `A B->C` to `C B->A`.

Comment (ilyagr) [reply lib/src/merge.rs:129]: Exactly. My example just has more sides. Adding an extra side to reordering from `A B->C` to `C B->A` can give you a reordering from `D E->A B->C` to `D E->C B->A`.

This might look more natural if you think of going from `∅➜A B➜C` to `∅➜D E➜A B➜C`. Note that if `D=E`, the two would be equivalent conflicts (so, you can add a side without changing the conflict, and then change one of the sides to get a non-trivial example with more sides_.
